### ISSUE TITLE
Corner-case: don't select first match when input changes because of completion

### ISSUE THREAD

**Issue by tsdh** [abo-abo/swiper#258]
Now the selection of the first candidate works after the input changed works properly as per #253.  But now I've found a corner-case where this behavior is not very intuitive.  Say you complete files and have the input "a" and candidates

```
aaab
aaac
```

Now I select the second candidate using `<down>` and hit `TAB` which is `ivy-partial-or-done`.  The result is that the input changes to `aaa` (good) but again the first candidate `aaab` is selected.

Well, in #253 I've suggested to perform the selection of the 1st candidate only if the last selected candidate isn't there anymore (point 2 of my issue description).  It would be very nice if that could be implemented.

**Comment by abo-abo** [comment]
Thanks. Just a reminder to myself, you and other people that raise issues, that it's always good to add a test to an issue if possible:

```elisp
(should (equal
         (ivy-with '(ivy-read "test" '("aaab" "aaac"))
                   "a C-n <tab> C-m")
         "aaac"))
```

This way, in the future it will be clear that this behavior must persist. And if anyone wonders why, running `git blame` on that test will bring up this issue discussion.

This is in contrast with, for example, bug#21644 with which I recently interacted. It's not clear what the behavior should be and why. The behavior is probably described in the manual, but it's not easy to locate.  It's also not easy to find the reasoning for why that was implemented in a particular way.
